tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts

```diff
@@ -0,0 +1,152 @@
+
+  beforeEach(() => {
+    const messageChannel = new MessageChannel();
+    channelTxSpy = spyOn(messageChannel.port1, 'postMessage');
+    spyOn(messageChannel.port2, 'postMessage').and.throwError(
+      'Should not call postMessage on port2'
+    );
+
+    spyOn(self, 'MessageChannel').and.returnValue(messageChannel);
+    workerPostMessageSpy = jasmine.createSpy('WorkerPool#postMessage');
+    spyOn(WorkerChart.workerPool, 'getNext').and.returnValue({
+      activeCount: 0,
+      postMessage: workerPostMessageSpy,
+      free: () => {},
+    });
+
+    onDrawEndSpy = jasmine.createSpy();
+    chart = new WorkerChart({
+      type: RendererType.WEBGL,
+      devicePixelRatio: 1,
+      callbacks: {onDrawEnd: onDrawEndSpy},
+      container: document.createElement('canvas'),
+      domDimension: {width: 100, height: 200},
+    });
+  });
+
+  it('initializes with canvas and ');
+
+  it('posts xScaleType message when setting xScaleType', () => {
+    chart.setXScaleType(ScaleType.LOG10);
+
+    expect(channelTxSpy).toHaveBeenCalledWith({
+      type: MainToGuestEvent.SCALE_UPDATE,
+      axis: 'x',
+      scaleType: ScaleType.LOG10,
+    });
+  });
+
+  it('posts yScaleType message when setting yScaleType', () => {
+    chart.setYScaleType(ScaleType.LOG10);
+
+    expect(channelTxSpy).toHaveBeenCalledWith({
+      type: MainToGuestEvent.SCALE_UPDATE,
+      axis: 'y',
+      scaleType: ScaleType.LOG10,
+    });
+  });
+
+  it('sends metadata', () => {
+    chart.setMetadata({
+      foo: {
+        id: 'foo',
+        displayName: 'Foo',
+        visible: false,
+        color: '#400',
+      },
+      bar: {
+        id: 'bar',
+        displayName: 'bar',
+        visible: true,
+        color: '#00f',
+      },
+    });
+
+    expect(channelTxSpy).toHaveBeenCalledWith({
+      type: MainToGuestEvent.SERIES_METADATA_CHANGED,
+      metadata: {
+        foo: {
+          id: 'foo',
+          displayName: 'Foo',
+          visible: false,
+          color: '#400',
+        },
+        bar: {
+          id: 'bar',
+          displayName: 'bar',
+          visible: true,
+          color: '#00f',
+        },
+      },
+    });
+  });
+
+  it('sends data in a compat form', () => {
+    chart.setData([
+      {
+        id: 'foo',
+        points: [
+          {x: 0, y: 0},
+          {x: 1, y: 1},
+        ],
+      },
+      {
+        id: 'bar',
+        points: [
+          {x: 0, y: -100},
+          {x: 100, y: 100},
+          {x: 200, y: -100},
+        ],
+      },
+    ]);
+
+    const expectedFlatttenedData = new Float32Array([
+      0,
+      0,
+      1,
+      1,
+      0,
+      -100,
+      100,
+      100,
+      200,
+      -100,
+    ]).buffer;
+    expect(channelTxSpy).toHaveBeenCalledWith(
+      {
+        type: MainToGuestEvent.SERIES_DATA_UPDATE,
+        idsAndLengths: [
+          {id: 'foo', length: 2},
+          {id: 'bar', length: 3},
+        ],
+        flattenedSeries: expectedFlatttenedData,
```

Comment:
I don’t object strongly to this, so I’ll leave it up to you, but perhaps
consider using `decompactDataSeries` here rather than poking at the
exact representation? Since the contract is “the host sends you
something that you can decompact to get the right data”, it’d be nice to
test against that specification rather than the current implementation
of the compaction format.